From a0943bd277ed9b5ecaf1c7c1ace7590b1b6da37d Mon Sep 17 00:00:00 2001
From: Shilin Victor <chrono.monochrome@gmail.com>
Date: Mon, 20 Aug 2018 22:57:43 +0300
Subject: [PATCH] am: Add NULL check in TaskRecord.getWindowContainerBounds

com.android.systemui process transact with Activity Manager in which
NullPointerException happens in TaskRecord.getWindowContainerBounds(),
then systemui will read this exception and so com.android.systemui
process force close.

It's hard to reproduce manually since it's monkey test issue, so what
can do is to check source codes, the NULL vlaue may be related with
TaskRecord.reparent() or ActivityStarter.setTaskFromIntentActivity()
or others, not sure the specific scenario, so currently adding NULL
check in this function to avoid crash for now.

diff --git a/services/core/java/com/android/server/am/TaskRecord.java b/services/core/java/com/android/server/am/TaskRecord.java
index 66fa001..0114644 100644
--- a/services/core/java/com/android/server/am/TaskRecord.java
+++ b/services/core/java/com/android/server/am/TaskRecord.java
@@ -573,7 +573,12 @@ void resizeWindowContainer() {
     }
 
     void getWindowContainerBounds(Rect bounds) {
-        mWindowContainerController.getBounds(bounds);
+        if (mWindowContainerController != null) {
+            mWindowContainerController.getBounds(bounds);
+            return;
+        }
+        Slog.w(TAG, "getWindowContainerBounds: task " + this + " has null window container.");
+        bounds.setEmpty();
     }
 
     /**
